Library/Homebrew/dev-cmd/update-python-resources.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require "cli/parser"
+
+module Homebrew
+  module_function
+
+  def update_python_resources_args
+    Homebrew::CLI::Parser.new do
+      usage_banner <<~EOS
+        `update-python-resources` [<options>] <formula>
+
+        Update versions for PyPI resource blocks in <formula>.
+      EOS
+      switch "-p", "--print-only",
+             description: "Print the updated resource blocks instead of changing the formula."
+      named :formula
+    end
+  end
+
+  def update_python_resources
+    update_python_resources_args.parse
+
+    formula = args.formulae.first
+    pypi_name = if formula.stable.url.start_with?("https://files.pythonhosted.org/packages/")
+      File.basename(formula.stable.url).match(/^(.+)-[a-z0-9.]+$/)[1]
+    else
+      formula.name
+    end
+
+    pypi_resources = formula.resources.select do |resource|
+      resource.url.start_with?("https://files.pythonhosted.org/packages/") && resource.name != "homebrew-virtualenv"
+    end
+
+    odie "No PyPI resources found in \"#{formula.name}\"" if pypi_resources.empty?
+
+    odie "\"python@3.8\" must be installed" unless Formula["python@3.8"].any_version_installed?
+
+    odie "\"pipgrip\" must be installed (`pip3 install pipgrip`)" unless File.exist?("#{HOMEBREW_PREFIX}/bin/pipgrip")
+
+    ohai "Retrieving PyPI dependencies for \"#{pypi_name}\"" unless args.print_only?
+    pipgrip_output = JSON.parse(Utils.safe_popen_read("#{HOMEBREW_PREFIX}/bin/pipgrip", "--json", pypi_name))
```

Comment:
> I realize we don't actually use resources for awscli since it drove maintainers insane c.f. [Homebrew/homebrew-core#5500](https://github.com/Homebrew/homebrew-core/pull/5500), but I couldn't think of another example.

This could be a success criteria: this tool should allow using `brew bump-formula-pr` to be used again for this formula.

Reply:
Makes sense. I'll add the version number and check for success before continuing

Reply:
The last example above sparked my curiosity! This error seems to be caused by [azure-cli==2.9.1 requiring release candidates in their (prod) releases](https://github.com/Azure/azure-cli/blob/azure-cli-2.9.1/src/azure-cli/requirements.py3.Linux.txt#L69), and pipgrip discarding such pre-releasea during available versions discovery for a package when the `--pre` flag isn't passed to pipgrip.

@jonchang if your last example is OK for pip (TODO check that pip will start downloading the downstream rc without passing `--pre` to pip), it should also be OK for pipgrip.

Reply:
Thanks for looking into this! It seems that `pip` does manage to install azure-cli==2.9.1 without needing `--pre`. Reading [PEP 440](https://www.python.org/dev/peps/pep-0440/#handling-of-pre-releases) it's clear why:

> Pre-releases of any kind, including developmental releases, are implicitly excluded from all version specifiers, unless they are already present on the system, explicitly requested by the user, or if the only available version that satisfies the version specifier is a pre-release.
> 
> By default, dependency resolution tools SHOULD:
> 
> *    accept already installed pre-releases for all version specifiers
> *    accept remotely available pre-releases for version specifiers where there is no final or post release that satisfies the version specifier
> *    exclude all other pre-releases from consideration

In this case, `pipgrep` should probably install 7.0.0rc1 as it applies to point 2 above (the pre-release being the only thing that satisfies the requirement).

Reply:
ref https://github.com/ddelange/pipgrip/issues/33

Reply:
Thanks, @ddelange! I've tested it locally and it has not thrown any errors.

@jonchang and @MikeMcQuaid Running `brew update-python-resources azure-cli` works without any errors. It does, however, make changes to the `azure-cli` formula. Next time `azure-cli` is updated we can see if `update-python-resources` works as expected for that formula.

(This, of course, won't work until https://github.com/Homebrew/homebrew-core/pull/58707 is completed)

Reply:
it was the combo of ~= (compatibility release clause) and a prerelease, plus no newer stable releases available to pypi. in this case azure-cli: https://github.com/Azure/azure-cli/blob/azure-cli-2.9.1/src/azure-cli/setup.py#L111